\documentclass[fleqn,10pt]{wlpeerj}
\usepackage{graphicx}
<<echo=FALSE,message=FALSE,results='hide'>>=
 #load required packages
 require(stringr)
 require(xtable)
 require(ggplot2)
@

<<echo=FALSE,message=FALSE,results='hide',cache=TRUE>>=
 #load all R analysis
 require(stringr)
 source("pub_script.R")
@

<<r setup, include=FALSE>>=
opts_chunk$set(dev = 'png')
opts_chunk$set(dpi=1000)
@
\title{Development of methods for using the pathoscope sequence analysis software to validate microbial test material purity.}

\author[1]{Nathan D. Olson}
\author[1]{Justin Zook}
\author[1]{Jayne Morrow}
\author[1]{Nancy Lin}
\affil[1]{Biosystems and Biomaterials Division, National Institute of Standards and Technology}

\keywords{Biodetection, Bioinformatics, Purity}

\begin{abstract}
Method validation for biodetection technologies requires high purity test materials.  Traditional methods for evaluating test material purity, e.g. polymerase chain reaction (PCR), require prior knowledge of the contaminant identity and have a limited level of detection (LOD).  Whole genome sequencing using next generation sequencing (NGS) technology addresses these limitations by providing a non-target specific approach and a lower LOD due to the large number of reads generated in a sequencing run \– ten thousand to over one million reads.  Here we present a novel application of the pathoscope sequence analysis program (http://sourceforge.net/projects/pathoscope/) coupled with an expanded reference dataset to evaluate test material purity and validate material identity using simulated whole genome sequencing datasets.  To evaluate the ability of pathoscope to identify contaminants in test materials, in silico generated datasets for \Sexpr{length(unique(single_matches$input_filename))} organisms from \Sexpr{length(unique(org_df$genus))} different genera \Sexpr{str_c(unique(org_df$genus)[-length(unique(org_df$genus))], collapse = ", ")} and \Sexpr{unique(org_df$genus)[length(unique(org_df$genus))]}).   First, the results from the analysis of the single organism datasets using pathoscope served as a baseline for the simulated contaminat analysis. Simulated datasets of 250 bp paried end reads with reads classified at the genus level provided suitable levels of specificity to identify potential test material contaminants.  To determine the method LOD, simulated reads from \Sexpr{length(unique(contam_matches$org2))} were mixed at known proportions. All contaminant organisms excluding B. anthracis amd F. tularensis were detected at 0.05 \% for all but 5 of the 29 the simulated test materials.  The detection levels for B. anthracis amd F. tularensis ranged between 5 \% down to 0.25 \%.  The approach presented here provides a sample independent, highly sensitive, and specific method for evaluating test material purity.  With the rapid decrease in NGS costs, this approach is a viable alternative to PCR based methods for evaluating test material purity and can be applied in a high throughput manner for screening panels of test materials.
\end{abstract}


\begin{document}

\flushbottom
\thispagestyle{empty}
 
\section*{Introduction}
Rapid, sensitive and accurate assays for detecting bacterial pathogens in food, water, clinical samples, and  suspicious biothreats is critical to public health and safety (REF).  Biodetection assays must be evaluated for assay sensitivity and specificity prior to deployment as well in the hands of the user to instill confidence in the actions made based on the assay results (\cite{Ieven2013,International2011,EPA2004,ISO/TS2010,Guide1998,Feldsine2002}).  Test materials are used to validate assay performance.  Test materials can be either purified cultures, genomic DNA or whole cells spiked into a matrix (\cite{EPA2004,ISO/TS2010,CLSI2010}).  Before being used to evaluate a biodetection assay the test material itself must be validated in terms of purity and identity to eliminate false positive results due to test material contaminants or false negative due to the test material being the wrong strain (\cite{CLSI2010}).  There are a number of potential sources of microbial contaminants  including the stock culture, the preservation medium, as well as airborne and laboratory contaminants (\cite{Marron2013,Shrestha2013,Tanner1998}).   
 
Current methods for evaluating test material purity include polymerase chain reaction (PCR) assays, metagenomics, and whole genome sequencing based approaches.  One PCR assay was developed to analyze protist cultures.  This PCR assay uses endpoint PCR for prokaryotes and eukaryotes with template dilutions (\cite{Marron2013}).  The benefit to PCR-based approaches is that they can be cost effective and fast if an applicable protocol exists.   However, PCR assays can only target specific contaminants.  While PCR assays can detect contaminants, this approach does not scale effectively for multiple contaminants and test materials. The bioinformatics tools developed to identify contaminants in metagenomic datasets, which include sequencing data from all organisms in a sample, can also be used to evaluate test material purity.  For example DeconSeq \cite{Schmieder2011} and a similar method QC-Chain (\cite{Zhou2013}) were developed to identify contaminants based on analysis of 16S ribosomal ribonucleic acid (rRNA) gene sequences or comparison of a subset of reads to a reference database using Basic local alignment search tool (BLAST).  Metagonomic-based methods are able to identify contaminants without any prior knowledge or assumptions regarding the identity of the organism(s).  However, methods based on 16S rRNA gene identification have limited resolution, as 16S rRNA sequences can only provide genus level taxonomic resolution at best.  Methods using BLAST-based searches represent a broader scale approach but are limited by the accuracy of the BLAST classification method.   The benefit to using metagenomic tools developed is that prior knowledge of the identity of the contaminant is not required; however this method is unable to identify contaminants to the strain level.   

Another approach to evaluating test material purity is through shotgun whole genome sequencing, sequence all DNA in a single organism sample.  A recently published bioinformatics method, pathoscope, was developed to detect pathogens and identify strains using whole genome sequencing data (\cite{Francis2013}).  This method benefits from the large sample size obtained using next generation sequencing for higher sensitivity and leverages algorithm advances for whole genome sequence mapping.  Mapping algorithms determine the optimal placement of reads relative to a reference sequence (\cite{Schbath2012}).  Reads are either uniquely or ambiguously mapped.  For uniquely mapped reads only a single optimal mapping location is identified, whereas for ambiguously mapped reads multiple optimal mapping locations are identified.  Pathoscope uses the number of reads that uniquely map to different genomes in the reference database to assign ambiguously mapped reads, reads that align equally well to multiple reference sequences.  The primary benefits to shotgun whole genome sequencing and subsequent pathoscope analysis approach are that prior knowledge of the contaminant is not required and it has the potential for higher sensitivity compared to other methods.  However, the main limitation to this method is the size of the reference database, namely that the genome of the contaminant or a closely related organism must be present in the database for the contaminant to be detected.  In this work, we developed a novel method to measure the purity of single organism test materials built upon the pathoscope software for pathogen detection.  This method is based on whole genome sequencing and utilizes pathoscope with an expanded reference database.  We will first present the specificity of the method using simulated data for single organisms.  Then, evaluate sensitivity of the method using simulated datasets generated to represent contaminated test material.  
 
\section*{Methods}
\label{sec:methods}
This bioinformatics approach to measuring test material purity consists of three steps: 1. Sequencing reads are generated for the test material using shotgun sequencing. 2. The reads are mapped to a reference database. For this step individual reads are aligned to all possible locations in the genomes in the reference database.   We generated a more exhaustive database compared to the pathoscope database including all prokaryotic organisms in the Genbank database (http://www.ncbi.nlm.nih.gov/genbank/, accessed 10/18/2013). 3. The resulting mapping file is processed with pathoscope to assess dataset composition (\cite{Francis2013}). For the mapping step the bwa mem mapping algorithm was used (http://bio-bwa.sourceforge.net).  This algorithm is capable of mapping reads to reference databases greater than 4 Gb in size.  The reference dataset included the chromosomes and plasmids for 2,709 bacterial genomes, including the strains used to validate the procedure.  Additionally, bwa mem can output all possible alignments for ambiguously mapped reads.  
 
To evaluate the specificity and sensitivity of the approach, simulated pure test material (single organism) as well as contaminated test material (mixtures of two organisms) were generated.  The organisms used in the study included common foodborne human pathogens and biothreat agents \ref{table:org-names}.  Simulated datasets were generated using the ART sequencing read simulator (\cite{Huang2012}).  The datasets were generated using the Illumina error models with paired end 75 and 250 base pair reads and 20 X mean coverage for each of the strains.

Simulated whole genome sequencing datasets for \Sexpr{ length(unique(single_matches$input_filename))} organisms from  \Sexpr{ length(unique(org_df$genus))} different genus (Table \ref{table:org-names}) were used to evaluate the method specificity. Datasets from the individual organisms were mapped to the all-bacterial genomes database using bwa mem.  The resulting mapping file was processed using pathoscope.   
   
<<genus-table, results='asis',echo=FALSE>>=
print(xtable(org_table2, caption="Strains used in validating method specificity and sensitivity.  The total strains is the number of strains for each genus used to validate the method specificity.  The representative strains were used to validate the method sensitivity. ", label = "table:org-names"),
      include.rownames=FALSE,
      caption.placement="top")
@
 
To evaluate the method sensitivity, simulated datasets with a test material strain and a contaminant strain were generated and evaluated.  Individual 250 bp paried end datasets for the test material and contaminant strains were generated as described in the specificity section with 20 X mean coverage.  The reads in the two datasets were subsampled in the following proportions: \Sexpr{proportions[-1]} and \Sexpr{proportions[length(proportions)]} where the first number represents the proportion of reads randomly selected from the test material dataset and the second number represents the proportion of reads randomly selected from the contaminant dataset.  For example, a proportion of 0.5:0.5, the combined dataset included a random subsampling of 50 \% of the reads from the target organism dataset and 50 \% of the reads from the contaminant organism dataset.  This apporach simulates the proportions of cells in a test material and not the amount of DNA, assuming unbiased DNA extraction.  Datasets were generated for all pairwise comparisons so that each of the 7 strains was both a contaminant and test material for all \Sexpr{length(proportions)} proportions combinations, for a total of \Sexpr{ length(unique(contam_matches$input_filename))}  datasets.  The individual organism datasets vary in the number of reads in the due to differences in genome size (table x?), such that 50 \% of one dataset may be different from the absolute number of reads than 50 \% of another dataset.   
 
To generate the simulated datasets, the single strain datasets were first mapped to the expanded reference database.  The full mapping files for the individual organisms were merged using samtools (http://samtools.sourceforge.net).  The resulting merged data files were downsampled to obtain the appropriate proportions of reads for each of the \Sexpr{length(unique(contam_matches$input_filename))} combinations using an in-house python script.  The resulting simulated contaminated datasets were processed using pathoscope, and the output files were analyzed to determine the limit of detection of the methods using the statistical programing language R \cite{R}.  All of the scripts used in the study are available at https://github.com/nate-d-olson/genomic-purity. 
 
\section*{Results and Discussion}
\subsection*{Specificity}    
The single organism datasets were used to assess method specificity.  Here specificity is defined as the methods ability to assign the reads to the appropriate organism.  
\begin{figure}
<<single-counts-plot, echo=FALSE, fig.height=3.5, fig.width=8>>=
ggplot(single_counts) + 
  geom_boxplot(aes(x = variable, y= value, color = size)) + 
  facet_grid(~target_genus) +
  scale_x_discrete(breaks = levels(single_counts$variable), 
                       labels = c("Genus", "Strain")) +
  labs(x = "Match Level", y = "Number of unqiue matches", color = "Size (bp)")+
  theme_bw() +
  theme(strip.text.x = element_text(face="italic"))
@
\caption{Number of unique matches for each genus by reads size and match level.  The black data points indicate outliers, the middle bars the median the boxes indicate the interquartile range (IQR, 25th to 75th percentiles) and the whiskers are 1.5 times the IQR. The plots are split on the x-axis by the simualted pure test material genus.}
\label{fig:single-counts-plot}
\end{figure}

\begin{figure}
<<single-specificity-plot, echo=FALSE>>=
ggplot(single_counts[single_counts$size == 250,]) + 
  geom_density(aes(x = value, fill = variable), alpha = 0.25) +
  scale_fill_discrete(breaks = levels(single_counts$variable), 
                      labels = c("Genus", "Strain")) +
  labs(x = "Numer of unique counts", 
       y = "Proportion of Total Counts", 
       fill = "Match Level") +
  theme_bw() +
  theme(legend.justification=c(1,0), legend.position=c(1,0.75))
@
\caption{Density distribution of the number of unqiue genus and strain level matches for the \Sexpr{length(unique(single_matches$input_filename))} simulated pure test material datasets.}
\label{fig:single-specificity-plot}
\end{figure}

For all organisms the number of both strain and genus matches decreases with increase in the length of the simulated reads \ref{fig:single-counts-plot}. The number of matches is dependent on the genetic diversity of the genomes for the genus in the database. Genus with greater diversity have fewer unique hits as fewer reads map incorectly to closely related genomes. The number of unique genus level matches for 250 bp simulated datasets provide a level of specificty allows for detection of contaminants in test materials with 99 \% of target organisms having \Sexpr{quantile(single_counts$value[single_counts$size == 250 & single_counts$variable == "genus_count"],probs=0.99)} or fewer unique genus level matches \ref{fig:single-specificity-plot}. 

\begin{figure}  
<<final-guess-plot, echo=FALSE, message=FALSE>>=
qplot(x = Final.Guess, data = single_matches, geom= "bar") + scale_x_log10() + theme_bw() + labs(x = "Estimated Proportions", y = "Abundance")
@
\caption{Distribution of the estimated proportion of reads in the datasets matching the individual sequences in the database for all simulated pure test material datasets.}
\label{fig:final-guess-plot}
\end{figure}

\begin{figure}
<<final-guess-exact-plot, echo=FALSE, message=FALSE>>=
#for exact matches only
#split by read size
#how to take into consideration plasmids vs reference genomes
qplot(x = Final.Guess, data = single_matches[single_matches$match == "exact",], geom= "bar") + scale_x_log10() + theme_bw() + labs(x = "Estimated Proportions", y = "Abundance")
@
\caption{Distribution of estimated proportion of reads in the datasets matching the correct reference sequence.}
\label{fig:final-guess-exact-plot}
\end{figure}

Only a few reads were mapped to a majority of the sequences in the database \ref{fig:final-guess-plot}. The estimated proportion of reads in the datasets mapped to a genome in the database 90 \% have an estimated proportion less than \Sexpr{ quantile(single_matches$Final.Guess, probs= 0.90)}.  Most of the reads are mapped to the match organisms genome in the reference database \ref{fig:final-guess-exact-plot}. Indicating that this method can be used to support claims regarding the identity of hte test material in addition to the test material purity.   This assumes that the test marerial genome sequence is in the database.  The authors of pathoscope test the algorithm performance on datasets where the test organims genome was not in the database and the algoirthm was able to accuractely identify the nearest genome in the reference database \cite{Francis2013}.  

\subsection*{Sensitivity}
\begin{figure}
<<contam-genus-plot, echo=FALSE, fig.height = 4, fig.wigth = 8>>=
ggplot(contam_counts) + 
  geom_raster(aes(x = org2_name, y = match_genus, fill = match_genus), show_guide = F) + 
  facet_grid(~org1_name, scale = "free_x") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text = element_text(face = "italic"),
        strip.text.x = element_text(face="italic")) + 
  labs(x = "Contaminant Name", y= "Match Genus")
@
\caption{Genus level matches for all test material and contaminant strain combinations.  Unc. Enterobacteriaceae is an unclassified genus in the family Enterobacteriaceae, Enterobacteriaceae bacterium strain FGI 57 (Genbank taxonomy ID 693444). Plots are split on the x axis by the test material strains.}
\label{fig:contam-genus-plot}
\end{figure}

Method sensitivity was defined as the detection of the contaminant strain genus for 250 bp simulated paried end reads.  The limit of detection is defined as the lowest contaminant strain proportion detected for each test material contaminant strain combination. The specificity of the species and strain level evaluation of the matches for the 250 bp and all match levels for the 75 bp datasets was low and the method was unable to differentiate the noise from the contaminant signal.  To reduce the level of background noise hits to sequences for the single organism datasets for the test material strains were filtered from the contaminated dataset results. For all contaminated test material datasets only the test material organim and contaminants were detected for the 250 bp datasets when evaluated at the genus level excluding E. coli \ref{fig:contam-genus-plot}.  Different simulated datasets were used for the specficity and sensitity parts of the study.  The hits to the test material genus represent hits to different species in the genus than the species reads were mapped to for the single organism datasets.  For all of the E. coli datasets reads were also mapped to the closely related genus Shigella as well as Citrobacter and an unclassified Enterobateriaceae.  

\begin{figure}
<< contam-LOD-plot, echo=FALSE,fig.height = 4, fig.wigth = 8>>=
ggplot(contam_genus_match) + 
 geom_raster(aes(y= as.factor(prop2), x = org2_tid, fill = match_genus)) + 
 facet_grid(.~org1_name, scale = "free_x") +
 labs(y = "Contaminant Spike Level", 
      x = "Contaminant", 
      fill = "Contaminant Genus") +
 theme_bw() +
 theme(strip.text = element_text(face="italic"), legend.position = "bottom", legend.direction="horizontal",axis.text.x = element_blank())
@ 
\caption{Limit of detection for all test material and contaminant strain combinations.  The plots are split on the x-axis by the different test material strains.}
\label{fig:contam-LOD-plot}
\end{figure}

The method limit of detection ranged from contamination of 5 \% to 0.000025 \% of the sample \ref{fig:contam-LOD-plot}.  Most (31 out of 42) of test material and contaminant combinations having limit of detection of 0.25 \%.  F. tularensis and B. anthracis had consistently lower estimated limits of detection when spiked into other test materials.  The lower limit of detection is due to the lower geneic diversity of the two genus and therefore fewer reads in the datasets map uniquely to sequences for the genus.  For the F. tularensis, B. anthracis, and Y. pestis only 36, 153, and 470 reads uniquely mapped to the reference sequences for the simulated datasets.  

\section*{Conclusions}
The combination of whole genome sequencing and characterization of the datasets using pathoscope and a large reference database provides a viable method for characterizing test material purity.  While this method is unable to identify contamiants to the strain level it can easily be combined with other methods such as qPCR for strain level quantitative validation for specific contaminants. Future work will include characterization of the method performance for different read sizes and error profiles (e.g. Ion Torrent and Pacific Biosciences data).  Additionally, we are working on revising the the reference database to include other common contaminants such as human mitochondria genome.  While incorporating the whole human genome is computationally prohibitive the mitochondrial genome could be added to the database with minimal additional computational cost.  

\section*{Acknowledgments}
The authors would like to thanks Dr. Steven Lund for his assistance in developing the study.  The Department of Homeland Security (DHS) Science and Technology Directorate supported this work under the Interagency Agreement HSHQPM-12-X-00078 with the National Institute of Standards and Technology (NIST).  Opinions expressed in this paper are the authors’ and do not necessarily reflect the policies and views of DHS, NIST, or affiliated venues.  Certain commercial equipment, instruments, or materials are identified in this paper in order to specify the experimental procedure adequately.  Such identification is not intended to imply recommendations or endorsement by NIST, nor is it intended to imply that the materials or equipment identified are necessarily the best available for the purpose.  Official contribution of NIST; not subject to copyrights in USA.

\bibliography{purity}

\end{document}
