Summary of single organism data analysis
========================================================
```{r echo=FALSE, message=FALSE}
#loading required packages and sourcing files
library(ggplot2)
library(stringr)
library(reshape2)
library(plyr)
library(xtable)
source("file_locations.R")
source("match_tid.R")
library(NCBI2R)
sim_matches <- read.csv(str_c(path_results_directory,"/sim_single_matches.csv", sep=""), stringsAsFactor = F)
```
# Dataset description:
Paired end Illumina reads simulated using ART read simulator for the following organisms (Table 1).
```{r echo =FALSE}
org_tid <- unique(sim_matches$org_tid)
org_name <- c()
for(i in org_tid){
  org_name <- c(org_name,python.call("getNameByTaxid", i))
}
org_df <- data.frame(tid = org_tid, name = org_name)
```
```{r results='asis'}
print(xtable(org_df),type='html')
```

# Methods summary:
Simulated datasets were mapped to a reference database of NCBI GenBank microbial genomes (need to record verion/ date) using BWA mem algorithm (REF).  All top matches were retained.  Resulting bam file was processed with pathoscope (results). The analysis below is a summary of the output generated by pathoscope.

# Results
## Number of unique Match org hits per combination
```{r}
sim_unique_counts <- ddply(sim_matches, .(org_tid, size), summarize, count = length(unique(match_tid)))
sim_unique_counts$size <- as.factor(sim_unique_counts$size)
ggplot(sim_unique_counts) + 
  geom_boxplot(aes(x = size, y = count), color = "grey") +
  geom_point(aes(x = size, y = count, color = as.character(org_tid)), size = 4) + 
  geom_line(aes(x = as.numeric(size), y = count, color = as.character(org_tid))) +
  labs(x = "Read Size (bp)", y = "Number of hits to unique organisms", color = "Organism")+
  theme_bw()
```
#### Observations:
For all organisms the number of unqiue matches decreases with increase in the length of the simulated reads.     
Questions:   
1. How do the differences in number of unique matches compare?      
2. How does the ranks compare (who has the largest drop in matches with the increase in size)?   

### Quantitaive analysis
Evlauation of match proportion distribtions
```{r }
sim_match_sum <- ddply(sim_matches, .(org_tid, size, match), summarize, prop = sum(Final.Guess))
sim_match_sum$match <- factor(sim_match_sum$match, levels = c("exact", "species", "genus","family","order","class", "phylum","no match"))
ggplot(sim_match_sum) + geom_line(aes(x = as.numeric(match), y = prop, linetype = as.character(size))) +
  theme_bw() +
  theme(axis.text.x=element_text(angle=-90)) +
  scale_x_continuous(breaks = 1:length(levels(sim_match_sum$match)), labels = levels(sim_match_sum$match)) +
  labs(x = "Match Classification", y = "Match Proportion", linetype = "Read Size (bp)") + facet_wrap(~org_tid)
```

Do the results look better as a table or a graph?
```{r}
sim_match_sum$prop <- round(sim_match_sum$prop, 3)
sim_match_table <- dcast(sim_match_sum[sim_match_sum$prop > 0.000,],org_tid*size~match,value.var = "prop", fill = "")
```
```{r results='asis', echo=FALSE}
 library(xtable)
  print(xtable(sim_match_table),type='html')
```

## Conclusions   
As expected the quality of the matches increases significantly read size.  For 250 bp reads, contaminats are theoretically easily detected when the lowest common taxonomic level is genus or higher.