#################################################################################################################
##
## Summary: Score matches for simulated data 
## Date: 1/11/2014
## Author: Nate Olson
## Affiliation: National Institute for Standards and Technology
## Dependencies: output from gather_sim.R (simulated_mix.csv)
##  GenBank run info files for any datasets analyzed
##  searchered SRA database using GenomeTakr and Bacillus
## Output files: real_data_2.csv
## 
#################################################################################################################

####------------------------------
# OBJECTIVE
# score matches on taxonomic level
####------------------------------

####--------------
##
## required inputs
##
####--------------

## pathoscope matches
#--------------------
path_matches <- read.csv("simulated_mix.csv", stringsAsFactors = F)
path_matches$X <- NULL
# 'data.frame':  8190 obs. of  20 variables:
# $ Genome                       : chr  "gi|384546269|ref|NC_017337.1|" "gi|387779217|ref|NC_017349.1|" "gi|82749777|ref|NC_007622.1|" "gi|255961454|ref|NC_006570.2|" ...
# $ Final.Guess                  : num  0.997463 0.000821 0.000435 0.000361 0.000286 ...
# $ Final.Best.Hit               : num  0.997463 0.000821 0.000435 0.000361 0.000286 ...
# $ Final.Best.Hit.Read.Numbers  : int  80206 66 35 29 23 12 11 7 5 2 ...
# $ Final.High.Confidence.Hits   : num  0.997463 0.000821 0.000435 0.000361 0.000286 ...
# $ Final.Low.Confidence.Hits    : int  0 0 0 0 0 0 0 0 0 0 ...
# $ Initial.Guess                : num  0.997463 0.000821 0.000435 0.000361 0.000286 ...
# $ Initial.Best.Hit             : num  0.997463 0.000821 0.000435 0.000361 0.000286 ...
# $ Initial.Best.Hit.Read.Numbers: int  80206 66 35 29 23 12 11 7 5 2 ...
# $ Initial.High.Confidence.Hits : num  0.997463 0.000821 0.000435 0.000361 0.000286 ...
# $ Initial.Low.Confidence.Hits  : int  0 0 0 0 0 0 0 0 0 0 ...
# $ input_filename               : chr  "159689-250-57589-250" "159689-250-57589-250" "159689-250-57589-250" "159689-250-57589-250" ...
# $ Aligned_Reads                : int  80410 80410 80410 80410 80410 80410 80410 80410 80410 80410 ...
# $ Mapped_Genome                : int  22 22 22 22 22 22 22 22 22 22 ...
# $ size                         : int  250 250 250 250 250 250 250 250 250 250 ...
# $ org1                         : int  159689 159689 159689 159689 159689 159689 159689 159689 159689 159689 ...
# $ org2                         : int  57589 57589 57589 57589 57589 57589 57589 57589 57589 57589 ...
# $ prop1                        : num  1 1 1 1 1 1 1 1 1 1 ...
# $ prop2                        : num  1 1 1 1 1 1 1 1 1 1 ...

## org1 and org2 indicate the source organisms, numbers are source directory uid

## taxonomy of pathoscope match
#------------------------------
path_match_taxa <- read.csv("bac_all_taxa.csv", stringsAsFactor = FALSE)
path_match_taxa$X <- NULL
# 'data.frame':  1295 obs. of  7 variables:
#   $ X      : int  1 2 3 4 5 6 7 8 9 10 ...
# $ full_name    : chr  "Acaryochloris marina" "Acetobacter pasteurianus" "Acetobacterium woodii" "Acetohalobium arabaticum" ...
# $ class  : chr  NA "Alphaproteobacteria" "Clostridia" "Clostridia" ...
# $ genus  : chr  "Acaryochloris" "Acetobacter" "Acetobacterium" "Acetohalobium" ...
# $ order  : chr  "Chroococcales" "Rhodospirillales" "Clostridiales" "Halanaerobiales" ...
# $ phylum : chr  "Cyanobacteria" "Proteobacteria" "Firmicutes" "Firmicutes" ...
# $ species: chr  "Acaryochloris marina" "Acetobacter pasteurianus" "Acetobacterium woodii" "Acetohalobium arabaticum" ...
path_match_taxa <- rename(x = path_match_taxa,replace=c(".id"="full_name"))

## taxonomy read source
#----------------------
source_taxa <- read.csv("ref_taxa.csv", stringsAsFactors = FALSE)
source_taxa$X <- NULL
# 'data.frame':  2684 obs. of  9 variables:
#   $ X        : int  1 2 3 4 5 6 7 8 9 10 ...
# $ full_name: chr  "Acaryochloris marina MBIC11017" "Acetobacterium woodii DSM 1030" "Acetobacter pasteurianus 386B" "Acetobacter pasteurianus IFO 3283 01 42C" ...
# $ class    : chr  NA "Clostridia" "Alphaproteobacteria" "Alphaproteobacteria" ...
# $ family   : chr  NA "Eubacteriaceae" "Acetobacteraceae" "Acetobacteraceae" ...
# $ genus    : chr  "Acaryochloris" "Acetobacterium" "Acetobacter" "Acetobacter" ...
# $ order    : chr  "Chroococcales" "Clostridiales" "Rhodospirillales" "Rhodospirillales" ...
# $ phylum   : chr  "Cyanobacteria" "Firmicutes" "Proteobacteria" "Proteobacteria" ...
# $ species  : chr  "Acaryochloris marina" "Acetobacterium woodii" "Acetobacter pasteurianus" "Acetobacter pasteurianus" ...
# $ uid      : chr  "58167" "88073" "214433" "158377" ...

# compare taxnomy of match and source
# report highest level of match species > genus > ...

# input files

###############################################################################
## adding unique identifer to all_bac data file
all_bac <- data.frame(genomes = readLines("all_bac_genomes_names.txt"))

for(i in 1:nrow(all_bac)){
  uid <- source_taxa$uid[i] 
  all_bac$uid[grep(source_taxa$full_name[i], all_bac$genomes)] <- source_taxa$uid[i]
}

###
# Merging path matches with all_bac
all_bac$Genome <- str_replace(all_bac$genomes,">","")
all_bac$Genome <- str_replace(all_bac$Genome," .*","")
all_bac$genomes <-NULL
path_matches <- join(path_matches, all_bac)

# 'data.frame':  2684 obs. of  9 variables:
#   $ X        : int  1 2 3 4 5 6 7 8 9 10 ...
# $ full_name: chr  "Acaryochloris marina MBIC11017" "Acetobacterium woodii DSM 1030" "Acetobacter pasteurianus 386B" "Acetobacter pasteurianus IFO 3283 01 42C" ...
# $ class    : chr  NA "Clostridia" "Alphaproteobacteria" "Alphaproteobacteria" ...
# $ family   : chr  NA "Eubacteriaceae" "Acetobacteraceae" "Acetobacteraceae" ...
# $ genus    : chr  "Acaryochloris" "Acetobacterium" "Acetobacter" "Acetobacter" ...
# $ order    : chr  "Chroococcales" "Clostridiales" "Rhodospirillales" "Rhodospirillales" ...
# $ phylum   : chr  "Cyanobacteria" "Firmicutes" "Proteobacteria" "Proteobacteria" ...
# $ species  : chr  "Acaryochloris marina" "Acetobacterium woodii" "Acetobacter pasteurianus" "Acetobacter pasteurianus" ...
# $ uid      : chr  "58167" "88073" "214433" "158377" ...

org1_match <- source_taxa
colnames(org1_match) <- str_c("org1_",colnames(org1_match, ), sep = "")
org1_match <-rename(org1_match, replace = c("org1_uid" = "org1"))
org2_match <- source_taxa
colnames(org2_match) <- str_c("org2_",colnames(org2_match, ), sep = "")
org2_match <-rename(org2_match, replace = c("org2_uid" = "org2"))


path_matches <- join(path_matches, source_taxa)
path_matches <- join(path_matches, org1_match)
path_matches <- join(path_matches, org2_match)


path_matches$org1_match[path_matches$phylum == path_matches$org1_phylum] <- "phylum"
path_matches$org1_match[path_matches$class == path_matches$org1_class] <- "class"
path_matches$org1_match[path_matches$order == path_matches$org1_order] <- "order"
path_matches$org1_match[path_matches$genus == path_matches$org1_genus] <- "genus"
path_matches$org1_match[path_matches$species == path_matches$org1_species] <- "species"
path_matches$org2_match[path_matches$phylum == path_matches$org2_phylum] <- "phylum"
path_matches$org2_match[path_matches$class == path_matches$org2_class] <- "class"
path_matches$org2_match[path_matches$order == path_matches$org2_order] <- "order"
path_matches$org2_match[path_matches$genus == path_matches$org2_genus] <- "genus"
path_matches$org2_match[path_matches$species == path_matches$org2_species] <- "species"


path_matches$org1_match[path_matches$uid == path_matches$org1] <- "strain"
path_matches$org2_match[path_matches$uid == path_matches$org2] <- "strain"
path_matches$org1_match[is.na(path_matches$org1_match)] <- "no match"
path_matches$org2_match[is.na(path_matches$org2_match)] <- "no match"

#need to expand to include higher level matches
#need to look into the robustness to the matching approach
#best approach maybe to work with directory names and fasta seq names
#fasta seq has unique id pathoscope returns, directory name has unique id
write.csv(path_matches,"sim_path_matches.csv", row.names = F)
