Summary of single organism data analysis
========================================================
```{r echo=FALSE, message=FALSE}
#loading required packages and sourcing files
library(ggplot2)
library(stringr)
library(reshape2)
library(plyr)
library(xtable)
source("purity_functions.R")
source("file_locations.R")
source("match_tid.R")
#sim_matches <- read.csv(str_c(path_results_directory,"/sim_single_matches.csv", sep=""), stringsAsFactor = F)
single_matches <- str_c(path_results_directory,c("/sim_single_matches.csv", "/sim_BA_matches.csv","/sim_EC_matches.csv"), sep="")
sim_matches <- ldply(single_matches,.fun = read.csv, stringsAsFactors = F)
#need to double check to make sure no redundancy- i.e. the same org is not in any of the three files
```
# Dataset description:
Paired end Illumina reads simulated using ART read simulator for the following organisms (Table 1).
```{r echo=FALSE, message=FALSE}
org_tid <- unique(sim_matches$org_tid)
org_name <- c()
#org_species <- c()
org_genus <- c()
for(i in org_tid){
  
  org_name <- c(org_name,python.call("getNameByTaxid", i))
  org_taxa <- GetTaxInfoLocal(i)
  org_taxa <- org_taxa[org_taxa$tid != 131567,]
  #org_species <- c(org_species, as.character(org_taxa$name[org_taxa$rank == "species"][1]))
  org_genus <- c(org_genus, as.character(org_taxa$name[org_taxa$rank == "genus"][1]))
}
org_df <- data.frame(org_tid, name = org_name, genus = org_genus) #species = org_species, 
org_table <- dcast(org_df,genus~.)
```
```{r results='asis', echo=FALSE}
print(xtable(org_table),type='html')
```

# Methods summary:
Simulated datasets were mapped to a reference database of NCBI GenBank microbial genomes (need to record verion/ date) using BWA mem algorithm (REF).  All top matches were retained.  Resulting bam file was processed with pathoscope (results). The analysis below is a summary of the output generated by pathoscope.

# Results
## Number of unique Match org hits per combination
```{r echo=FALSE, message=FALSE}
# sim_unique_counts <- ddply(sim_matches, .(org, org_tid, size), summarize, count = length(unique(match_tid)))
# sim_unique_counts$size <- as.factor(sim_unique_counts$size)
# sim_unique_counts <- join(sim_unique_counts, org_df)
# ggplot(sim_unique_counts) + 
#   geom_boxplot(aes(x = size, y = count), color = "grey") +
#   geom_point(aes(x = size, y = count, color = as.character(org_tid)), size = 4, alpha = 0.5, show_guide = F) + 
#   geom_line(aes(x = as.numeric(size), y = count, color = as.character(org_tid)), alpha = 0.5, show_guide = F) +
#   labs(x = "Read Size (bp)", y = "Number of hits to unique organisms", color = "Organism")+
#   theme_bw() +
#   facet_wrap(~genus)
# for(i in c(0.1,0.01,0.001,0.0001, 0)){
#   print(i)
#   print(unique_count_plot(sim_matches[sim_matches$Final.Guess > i,]))
# }
unique_count_plot(sim_matches)
```
#### Observations:
For all organisms the number of unqiue matches decreases with increase in the length of the simulated reads. 
The number of matches is dependent on the genetic diversity of the genus/ species for the target organism.

## Predicting a base line level for contaminat detection.
Overall distribition of Final.Guess values.  Final.Guess is the proprotion of the total number of aligned reads to the number of a reads assinged to a genome in the reference database.
```{r echo=FALSE, message=FALSE}
qplot(x = Final.Guess, data = sim_matches, geom= "bar") + scale_x_log10() + theme_bw()
```
The 90 % of the hits have a Final.Guess values of less than `r quantile(sim_matches$Final.Guess, probs= 0.90)`.  

```{r echo=FALSE, message=FALSE}
sim_matches <- join(sim_matches, org_df)
sim_matches$match <- factor(sim_matches$match, levels = c("exact", "species", "genus","family","order","class", "phylum","no match"))
```
The proportion of the values for Final.Guess is dependent on the match level and the genus of the target organims.  
Table of match genus vs. match hit level    
```{r echo=FALSE, message=FALSE, fig.width=12}
sim_quant <- ddply(sim_matches, .(genus,match,size),summarize, p90 = quantile(Final.Guess, probs=0.9))
ggplot(sim_quant) + geom_line(aes(x = as.numeric(match), y = p90, color = genus)) + scale_y_log10() + scale_x_continuous(breaks = 1:length(levels(sim_quant$match)), labels = levels(sim_quant$match)) + labs(y = "Final.Guess value 90th Percentile", x = "Shared Taxonomic Level with Match") + facet_wrap(~size) + theme_bw()
```

```{r echo=FALSE, message=FALSE} 
#results='asis', 
 library(xtable)
 # print(xtable(dcast(sim_quant,genus*size~match, value.var = "p90")),type='html')
```

## Conclusions   
As expected the quality of the matches increases significantly read size.  For 250 bp reads, contaminats are theoretically easily detected when the lowest common taxonomic level is genus or higher.  
Include statement on how these results can be used to make preliminary conclusions about identification.  The top match for all organisms was the exact match to the target.